library/Booster/Pattern/ApplyEquations.hs

```diff
@@ -688,15 +681,15 @@ simplifyConstraint ::
     Predicate ->
     io (Either EquationFailure Predicate, [EquationTrace], SimplifierCache)
 simplifyConstraint doTracing def mbApi cache p =
-    runEquationT doTracing def mbApi cache $ simplifyConstraint' p
+    runEquationT doTracing def mbApi cache $ simplifyConstraint' True p
 
 -- version for internal nested evaluation
-simplifyConstraint' :: MonadLoggerIO io => Predicate -> EquationT io Predicate
+simplifyConstraint' :: MonadLoggerIO io => Bool -> Predicate -> EquationT io Predicate
 -- We are assuming all predicates are of the form 'true \equals P' and
 -- evaluating them using simplifyBool if they are concrete.
 -- Non-concrete \equals predicates are simplified using evaluateTerm.
-simplifyConstraint' = \case
-    EqualsTerm TrueBool t@(Term attributes _)
+simplifyConstraint' recurse = \case
+    p@(EqualsTerm TrueBool t@(Term attributes _))
```

Comment:
The `recurse` flag might do for a short while but we won't have the behaviour from before. Simplify requests (which call `simplifyConstraint`) can now go into loops.
IMHO we should add loop protection and a recursion limit. I already had that code in #352  (see the modified `EquationState` there)

Reply:
Can you test that change on jevm/kontrol proofs? As long as the timings look OK I'm happy to do this

Reply:
Sure, I'll run the tests on my branch. 
I like your refactoring, we should by all means keep it.

Reply:
As discussed, withdrawing the comment about recursion :brain: :tornado: